integration-tests/tests/src/tests/counter.ts

```diff
@@ -0,0 +1,994 @@
+        const objectWithInt = this.realm.create<IWithRegularInt>(WithRegularIntSchema.name, {
+          int: 10,
+        });
+        // Create and object with a counter that will be used for setting an 'int' property.
+        const { counter } = this.realm.create<IWithCounter>(WithCounterSchema.name, {
+          counter: 20,
+        });
+        return { objectWithInt, counter };
+      });
+      expectCounter(counter);
+      expect(counter.value).equals(20);
+      expect(objectWithInt.int).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          // @ts-expect-error Testing incorrect type.
+          objectWithInt.int = counter;
+        });
+      }).to.throw("Counters can only be used when 'counter' is declared in the property schema");
+      expect(objectWithInt.int).equals(10);
+    });
+
+    it("throws when getting the count on an invalidated obj", function (this: RealmContext) {
+      const object = this.realm.write(() => {
+        return this.realm.create<IWithCounter>(WithCounterSchema.name, {
+          counter: 10,
+        });
+      });
+      const counter = object.counter;
+      expectCounter(counter);
+      expect(counter.value).equals(10);
+      expect(this.realm.objects(WithCounterSchema.name).length).equals(1);
+
+      this.realm.write(() => {
+        this.realm.delete(object);
+      });
+      expect(this.realm.objects(WithCounterSchema.name).length).equals(0);
+
+      expect(() => counter.value).to.throw("Accessing object which has been invalidated or deleted");
+    });
+
+    it("throws when setting a mixed to a counter via object creation", function (this: RealmContext) {
+      const { counter } = this.realm.write(() => {
+        return this.realm.create<IWithCounter>(WithCounterSchema.name, {
+          counter: 10,
+        });
+      });
+      expectCounter(counter);
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          this.realm.create<IWithMixed>(WithMixedSchema.name, { mixed: counter });
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
+      expect(this.realm.objects(WithMixedSchema.name).length).equals(0);
+    });
+
+    it("throws when setting a mixed to a counter via object setter", function (this: RealmContext) {
+      const { counter, objectWithMixed } = this.realm.write(() => {
+        const counter = this.realm.create<IWithCounter>(WithCounterSchema.name, {
+          counter: 10,
+        }).counter;
+        const objectWithMixed = this.realm.create<IWithMixed>(WithMixedSchema.name, { mixed: 20 });
+
+        return { counter, objectWithMixed };
+      });
+      expectCounter(counter);
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          objectWithMixed.mixed = counter;
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
```

Comment:
Perhaps we should capture an issue to track lifting this restriction in the future?

Reply:
I think it's more a server-side restriction at the moment, with not being able to merge it correctly. @nirinchev did you know the specific restriction here?